partner_firstname/models.py

```diff
@@ -80,13 +83,20 @@ def _inverse_name(self):
         # Company name goes to the lastname
         if self.is_company or self.name is False:
             parts = [self.name, False]
-
         # Guess name splitting
         else:
-            parts = self.name.split(" ", 1)
-            while len(parts) < 2:
-                parts.append(False)
-
+            parts = self.name.split(" ")
+            if len(parts) > 1:
+                if self.company_id.names_order == 'first_last':
+                    first = parts[0]
+                    last = u" ".join(parts[1:])
+                else:
+                    first = parts[-1]
+                    last = u" ".join(parts[:-1])
+                parts = [last, first]
+            else:
+                while len(parts) < 2:
+                    parts.append(False)
         self.lastname, self.firstname = parts
```

Comment:
It's an undocumented feature. It's the same as [`write()`](https://www.odoo.com/documentation/8.0/reference/orm.html#openerp.models.Model.write), but works in temporary enviroments (onchange). Read https://github.com/odoo/odoo/issues/7417 and you'll understand.